python/ray/train/huggingface/accelerate/_accelerate_utils.py

```diff
@@ -0,0 +1,172 @@
+        self.config_raw = config_raw
+        self.deepspeed_config_raw = deepspeed_config_raw
+
+    def __bool__(self) -> bool:
+        return bool(self.config_raw)
+
+
+def simple_launcher(args):
+    _, current_env = prepare_simple_launcher_cmd_env(args)
+
+    os.environ.update(current_env)
+
+
+def multi_gpu_launcher(args):
+    current_env = prepare_multi_gpu_env(args)
+    os.environ.update(current_env)
+
+
+def deepspeed_launcher(args):
+    if not is_deepspeed_available():
+        raise ImportError(
+            "DeepSpeed is not installed => run `pip3 install deepspeed` "
+            "or build it from source."
+        )
+
+    _, current_env = prepare_deepspeed_cmd_env(args)
+
+    os.environ.update(current_env)
+
+
+def launch_command(args):
+    args, defaults, mp_from_config_flag = _validate_launch_command(args)
+
+    # Use the proper launcher
+    if args.use_deepspeed and not args.cpu:
+        args.deepspeed_fields_from_accelerate_config = (
+            list(defaults.deepspeed_config.keys()) if defaults else []
+        )
+        if mp_from_config_flag:
+            args.deepspeed_fields_from_accelerate_config.append("mixed_precision")
+        args.deepspeed_fields_from_accelerate_config = ",".join(
+            args.deepspeed_fields_from_accelerate_config
+        )
+        deepspeed_launcher(args)
+    elif args.use_fsdp and not args.cpu:
+        multi_gpu_launcher(args)
+    elif args.use_megatron_lm and not args.cpu:
+        multi_gpu_launcher(args)
+    elif args.multi_gpu and not args.cpu:
+        multi_gpu_launcher(args)
+    elif args.tpu and not args.cpu:
+        raise NotImplementedError("")
+    elif (
+        defaults is not None
+        and defaults.compute_environment == ComputeEnvironment.AMAZON_SAGEMAKER
+    ):
+        raise NotImplementedError(
+            "Cannot use SageMaker compute environment with Ray Train."
+        )
+    else:
+        simple_launcher(args)
+
+
+def load_accelerate_config(
+    accelerate_config: Optional[Union[str, dict]],
+) -> Tuple[str, Optional[str]]:
+    if isinstance(accelerate_config, dict):
+        ctx = tempfile.TemporaryDirectory
+    else:
+        ctx = nullcontext
+
+    with ctx() as tempdir:
+        if isinstance(accelerate_config, dict):
+            tempdir = Path(tempdir)
+            accelerate_config_path = str(tempdir / "accelerate_config.json")
+            accelerate_config.setdefault("num_processes", 1)
+            accelerate_config.setdefault("distributed_type", "NO")
+            accelerate_config.setdefault("mixed_precision", "no")
+            accelerate_config.setdefault("compute_environment", "LOCAL_MACHINE")
+            accelerate_config.setdefault("use_cpu", True)
+            config = ClusterConfig(**accelerate_config)
+            config.to_json_file(accelerate_config_path)
+        else:
+            accelerate_config_path = (
+                str(accelerate_config) if accelerate_config else default_config_file
+            )
```

Comment:
wait, if we get a dict, we first write it into a file, then load it back in right after?
why don't we keep the dict, and load the file only if accelerate_config is a string path?

Reply:
yes, that way we ensure that Accelerate works correctly (as it does some side effects on loading). You can't pass a dict in Accelerate itself, which is why we need to do this dance with files - otherwise we would have to use Accelerate private APIs to replicate the behavior

Reply:
I see. can you add this explanation as comments in the file? key piece of information for anyone who needs to maintain this code.

Reply:
Added